// @effect/cluster/dist/cjs/HttpRunner.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.toHttpEffectWebsocket = exports.toHttpEffect = exports.layerWebsocketOptions = exports.layerWebsocketClientOnly = exports.layerWebsocket = exports.layerHttpOptions = exports.layerHttpClientOnly = exports.layerHttp = exports.layerClientProtocolWebsocketDefault = exports.layerClientProtocolWebsocket = exports.layerClientProtocolHttpDefault = exports.layerClientProtocolHttp = exports.layerClient = void 0;
var HttpClient = _interopRequireWildcard(require("@effect/platform/HttpClient"));
var HttpClientRequest = _interopRequireWildcard(require("@effect/platform/HttpClientRequest"));
var HttpRouter = _interopRequireWildcard(require("@effect/platform/HttpLayerRouter"));
var Socket = _interopRequireWildcard(require("@effect/platform/Socket"));
var RpcClient = _interopRequireWildcard(require("@effect/rpc/RpcClient"));
var RpcSerialization = _interopRequireWildcard(require("@effect/rpc/RpcSerialization"));
var RpcServer = _interopRequireWildcard(require("@effect/rpc/RpcServer"));
var Effect = _interopRequireWildcard(require("effect/Effect"));
var Layer = _interopRequireWildcard(require("effect/Layer"));
var _Runners = _interopRequireWildcard(require("./Runners.js"));
var Runners = _Runners;
var RunnerServer = _interopRequireWildcard(require("./RunnerServer.js"));
var Sharding = _interopRequireWildcard(require("./Sharding.js"));
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
/**
 * @since 1.0.0
 */

/**
 * @since 1.0.0
 * @category Layers
 */
const layerClientProtocolHttp = options => Layer.effect(_Runners.RpcClientProtocol)(Effect.gen(function* () {
  const serialization = yield* RpcSerialization.RpcSerialization;
  const client = yield* HttpClient.HttpClient;
  const https = options.https ?? false;
  return address => {
    const clientWithUrl = HttpClient.mapRequest(client, HttpClientRequest.prependUrl(`http${https ? "s" : ""}://${address.host}:${address.port}/${options.path}`));
    return RpcClient.makeProtocolHttp(clientWithUrl).pipe(Effect.provideService(RpcSerialization.RpcSerialization, serialization));
  };
}));
/**
 * @since 1.0.0
 * @category Layers
 */
exports.layerClientProtocolHttp = layerClientProtocolHttp;
const layerClientProtocolHttpDefault = exports.layerClientProtocolHttpDefault = /*#__PURE__*/layerClientProtocolHttp({
  path: "/"
});
/**
 * @since 1.0.0
 * @category Layers
 */
const layerClientProtocolWebsocket = options => Layer.effect(_Runners.RpcClientProtocol)(Effect.gen(function* () {
  const serialization = yield* RpcSerialization.RpcSerialization;
  const https = options.https ?? false;
  const constructor = yield* Socket.WebSocketConstructor;
  return Effect.fnUntraced(function* (address) {
    const socket = yield* Socket.makeWebSocket(`ws${https ? "s" : ""}://${address.host}:${address.port}/${options.path}`).pipe(Effect.provideService(Socket.WebSocketConstructor, constructor));
    return yield* RpcClient.makeProtocolSocket().pipe(Effect.provideService(Socket.Socket, socket), Effect.provideService(RpcSerialization.RpcSerialization, serialization));
  });
}));
/**
 * @since 1.0.0
 * @category Layers
 */
exports.layerClientProtocolWebsocket = layerClientProtocolWebsocket;
const layerClientProtocolWebsocketDefault = exports.layerClientProtocolWebsocketDefault = /*#__PURE__*/layerClientProtocolWebsocket({
  path: "/"
});
/**
 * @since 1.0.0
 * @category Http App
 */
const toHttpEffect = exports.toHttpEffect = /*#__PURE__*/Effect.gen(function* () {
  const handlers = yield* Layer.build(RunnerServer.layerHandlers);
  return yield* RpcServer.toHttpApp(Runners.Rpcs, {
    spanPrefix: "RunnerServer",
    disableTracing: true
  }).pipe(Effect.provide(handlers));
});
/**
 * @since 1.0.0
 * @category Http App
 */
const toHttpEffectWebsocket = exports.toHttpEffectWebsocket = /*#__PURE__*/Effect.gen(function* () {
  const handlers = yield* Layer.build(RunnerServer.layerHandlers);
  return yield* RpcServer.toHttpAppWebsocket(Runners.Rpcs, {
    spanPrefix: "RunnerServer",
    disableTracing: true
  }).pipe(Effect.provide(handlers));
});
/**
 * @since 1.0.0
 * @category Layers
 */
const layerClient = exports.layerClient = /*#__PURE__*/Sharding.layer.pipe(/*#__PURE__*/Layer.provideMerge(Runners.layerRpc));
/**
 * A HTTP layer for the `Runners` services, that adds a route to the provided
 * `HttpRouter`.
 *
 * @since 1.0.0
 * @category Layers
 */
const layerHttpOptions = options => RunnerServer.layerWithClients.pipe(Layer.provide(RpcServer.layerProtocolHttp(options)));
/**
 * @since 1.0.0
 * @category Layers
 */
exports.layerHttpOptions = layerHttpOptions;
const layerWebsocketOptions = options => RunnerServer.layerWithClients.pipe(Layer.provide(RpcServer.layerProtocolWebsocket(options)));
/**
 * @since 1.0.0
 * @category Layers
 */
exports.layerWebsocketOptions = layerWebsocketOptions;
const layerHttp = exports.layerHttp = /*#__PURE__*/HttpRouter.serve(layerHttpOptions({
  path: "/"
})).pipe(/*#__PURE__*/Layer.provide(layerClientProtocolHttpDefault));
/**
 * @since 1.0.0
 * @category Layers
 */
const layerHttpClientOnly = exports.layerHttpClientOnly = /*#__PURE__*/RunnerServer.layerClientOnly.pipe(/*#__PURE__*/Layer.provide(layerClientProtocolHttpDefault));
/**
 * @since 1.0.0
 * @category Layers
 */
const layerWebsocket = exports.layerWebsocket = /*#__PURE__*/HttpRouter.serve(layerWebsocketOptions({
  path: "/"
})).pipe(/*#__PURE__*/Layer.provide(layerClientProtocolWebsocketDefault));
/**
 * @since 1.0.0
 * @category Layers
 */
const layerWebsocketClientOnly = exports.layerWebsocketClientOnly = /*#__PURE__*/RunnerServer.layerClientOnly.pipe(/*#__PURE__*/Layer.provide(layerClientProtocolWebsocketDefault));
